broker/src/bin/neon_broker.rs

```diff
@@ -0,0 +1,515 @@
+                chan_to_timeline_sub.num_subscribers -= 1;
+                if chan_to_timeline_sub.num_subscribers == 0 {
+                    self.chans_to_timeline_subs.remove(&ttid);
+                }
+            }
+        }
+    }
+}
+
+// SharedState wrapper.
+#[derive(Clone)]
+struct Registry {
+    shared_state: Arc<RwLock<SharedState>>,
+    chan_size: usize,
+}
+
+impl Registry {
+    // Register new publisher in shared state.
+    pub fn register_publisher(&self) -> Publisher {
+        let publisher = self.shared_state.write().register_publisher(self.clone());
+        trace!("registered publisher {}", publisher.id);
+        publisher
+    }
+
+    pub fn unregister_publisher(&self, publisher: &Publisher) {
+        self.shared_state.write().unregister_publisher();
+        trace!("unregistered publisher {}", publisher.id);
+    }
+
+    // Register new subscriber in shared state.
+    pub fn register_subscriber(&self, sub_key: SubscriptionKey) -> Subscriber {
+        let subscriber =
+            self.shared_state
+                .write()
+                .register_subscriber(sub_key, self.clone(), self.chan_size);
+        trace!("registered subscriber {}", subscriber.id);
+        subscriber
+    }
+
+    // Unregister the subscriber
+    pub fn unregister_subscriber(&self, subscriber: &Subscriber) {
+        self.shared_state
+            .write()
+            .unregister_subscriber(subscriber.key);
+        trace!("unregistered subscriber {}", subscriber.id);
+    }
+}
+
+// Private subscriber state.
+struct Subscriber {
+    id: SubId,
+    key: SubscriptionKey,
+    // Subscriber receives messages from publishers here.
+    sub_rx: broadcast::Receiver<SafekeeperTimelineInfo>,
+    // to unregister itself from shared state in Drop
+    registry: Registry,
+}
+
+impl Drop for Subscriber {
+    fn drop(&mut self) {
+        self.registry.unregister_subscriber(self);
+    }
+}
+
+// Private publisher state
+struct Publisher {
+    id: PubId,
+    registry: Registry,
+}
+
+impl Publisher {
+    // Send msg to relevant subscribers.
+    pub fn send_msg(&mut self, msg: &SafekeeperTimelineInfo) -> Result<(), Status> {
+        // send message to subscribers for everything
+        let shared_state = self.registry.shared_state.read();
+        // Err means there is no subscribers, it is fine.
+        shared_state.chan_to_all_subs.send(msg.clone()).ok();
+
+        // send message to per timeline subscribers
+        let ttid =
+            parse_proto_ttid(msg.tenant_timeline_id.as_ref().ok_or_else(|| {
+                Status::new(Code::InvalidArgument, "missing tenant_timeline_id")
+            })?)?;
+        if let Some(subs) = shared_state.chans_to_timeline_subs.get(&ttid) {
+            // Err can't happen here, as rx is destroyed after removing from the
+            // map the last subscriber.
+            subs.chan.send(msg.clone()).unwrap();
```

Comment:
Lets use `.expect(...)` here instead of unwrap

Reply:
done